drafts/str4d-overwinter-sighash/draft1.rst

```diff
@@ -0,0 +1,279 @@
+  case that a wrong value is provided and signed, the signature would be invalid and no funds would be lost.
+  [#offline-wallets]_
+
+Deploying the aforementioned fixes in the original script system is not a simple task.
+
+
+Specification
+=============
+
+A new transaction digest algorithm is defined, but only applicable from the Overwinter upgrade block height
+[#ZIP0000]_::
+  [TODO: Pick one] BLAKE2[b-256|s] of the serialization of:
+    1. nVersion of the transaction (4-byte little endian)
+    2. hashPrevouts (32-byte hash)
+    3. hashSequence (32-byte hash)
+    4. hashOutputs (32-byte hash)
+    5. hashJoinSplits (32-byte hash)
+    6. nLocktime of the transaction (8-byte little endian)
+    7. sighash type of the signature (4-byte little endian)
+    8. If we are serializing an input (ie. this is not a JoinSplit signature hash):
+       a. outpoint (32-byte hash + 4-byte little endian) 
+       b. scriptCode of the input (serialized as scripts inside CTxOuts) [TODO]
+       c. value of the output spent by this input (8-byte little endian)
+       d. nSequence of the input (4-byte little endian)
+
+The BLAKE2[b-256|s] personalization field will be set to::
+
+  ["ZcashSigHash"|"ZcSH"] || BRANCH_ID
+
+Semantics of the original sighash types remain unchanged, except the followings:
+
+#. The way of serialization is changed;
+
+#. All sighash types commit to the amount being spent by the signed input;
+
+#. ``SINGLE`` does not commit to the input index. When ``ANYONECANPAY`` is not set, the semantics are
+   unchanged since ``hashPrevouts`` and ``outpoint`` together implictly commit to the input index. When
+   ``SINGLE`` is used with ``ANYONECANPAY``, omission of the index commitment allows permutation of the
+   input-output pairs, as long as each pair is located at an equivalent index.
+
+Field definitions
+-----------------
+
+The items 1, 6, 7, 8a, 8d have the same meaning as the original algorithm. [#wiki-checksig]_
+
+2: ``hashPrevouts``
+```````````````````
+* If the ``ANYONECANPAY`` flag is not set, ``hashPrevouts`` is the double SHA256 of the serialization of all
+  input outpoints;
+
+* Otherwise, ``hashPrevouts`` is a ``uint256`` of ``0x0000......0000``.
+
+3: ``hashSequence``
+```````````````````
+* If none of the ``ANYONECANPAY``, ``SINGLE``, ``NONE`` sighash type is set, ``hashSequence`` is the double
+  SHA256 of the serialization of ``nSequence`` of all inputs;
+
+* Otherwise, ``hashSequence`` is a ``uint256`` of ``0x0000......0000``.
+
+4: ``hashOutputs``
+``````````````````
+* If the sighash type is neither ``SINGLE`` nor ``NONE``, ``hashOutputs`` is the double SHA256 of the
+  serialization of all output amount (8-byte little endian) with ``scriptPubKey`` (serialized as scripts
+  inside CTxOuts);
+
+* If sighash type is ``SINGLE`` and the input index is smaller than the number of outputs, ``hashOutputs`` is
+  the double SHA256 of the output amount with ``scriptPubKey`` of the same index as the input;
```

Comment:
(in reply to https://github.com/str4d/zips/commit/3d2923933d27648c8e2e6c2c0dfbfe511aded537#r26723147)

@daira agreed, the wording from BIP 143 isn't clear. It means the same serialization as the first case, but only of the output corresponding to the input signature being created (instead of the whole `vout`).